File: Javascript/src/01.AlgorithmicToolbox/Week3GreedyAlgorithms/largestNumber.js
/**
 * Problem 6 - Maximum Salary
 * Compose the largest number from a sequence of positive integers
 * @param {Array.<{number}>} sequence sequence of numbers
 * @returns {number} largest number composable from the sequencea
 */
export function largestNumber(sequence) {
  const largestNumberString = sequence
    .sort((a, b) => {
      const ab = parseInt(`${a}${b}`, 10);
      const ba = parseInt(`${b}${a}`, 10);
      return ba - ab;
    })
    .join("");
  return parseInt(largestNumberString, 10);
}

File: Rust/src/_01_algorithmic_toolbox/week3_greedy_algorithms/largest_number.rs
/// Problem 6 - Maximum Salary
///
/// Compose the largest number from a sequence of positive integers
pub fn largest_number(mut sequence: Vec<u64>) -> u64 {
    sequence.sort_by(|a, b| {
        let ab = format!("{}{}", a, b);
        let ba = format!("{},{}", b, a);
        return ba.cmp(&ab);
    });

    sequence
        .into_iter()
        .map(|number| format!("{}", number))
        .collect::<Vec<std::string::String>>()
        .join("")
        .parse::<u64>()
        .unwrap()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn returns_maximum_composable_number_from_sequence() {
        assert_eq!(largest_number(vec![9, 4, 6, 1, 9]), 99641);
        assert_eq!(largest_number(vec![23, 39, 92]), 923923);
        assert_eq!(largest_number(vec![21, 2]), 221);
    }
}
